### PR TITLE
Update `_safe_divide` to allow `Accuracy` to run on the GPU

### LINKED ISSUE #2638: Accuracy incurs a GPU/CPU sync point
## 🐛 Bug

If running Accuracy inside the training loop, training is slowed due to CPU/GPU sync point

### To Reproduce

Steps to reproduce the behavior...

Pull Request includes reproduction, a unit test which fails before the modification and passes after.
[#1 Update _safe_divide to allow Accuracy to run on the GPU](https://github.com/ndrwrbgs/torchmetrics/pull/1)

_P.S. I opened it in the fork due to the question # 1 under "Before submitting", but [Contributing.md](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md) does not mandate discussion prior "Submit a PR!", would it be possible to update one or the other with the latest guidance?_

### Expected behavior

Accuracy can be calculated on the GPU, without syncing

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source): conda    torchmetrics              1.4.0.post0        pyhd8ed1ab_0    conda-forge
- Python & PyTorch Version (e.g., 1.0):  python                    3.11.9          hb806964_0_cpython    conda-forge
- Any other relevant information such as OS (e.g., Linux):

### PR DESCRIPTION
**Pull request by ndrwrbgs**
Caused a CPU/GPU sync point. torch.where can accept the float directly

## What does this PR do?

Fixes #2638

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>Ensure CPU/GPU sync point avoided for Accuracy metric</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2640.org.readthedocs.build/en/2640/

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Comment by ndrwrbgs** [comment]
Oh! Thank you @Borda for doing cleanups for me. I apologize I didn't get pre-commit installed until after the initial commit.

**Comment by Borda** [comment]
> Oh! Thank you @Borda for doing cleanups for me. I apologize I didn't get pre-commit installed until after the initial commit.

no worries :)

**Comment by ndrwrbgs** [comment]
Of course! I’m grateful for the good test coverage and can see what’s happening here.

Downgrading to draft so it’s off the docket until it is ready, I overestimated the ease.

**Comment by Borda** [comment]
> Downgrading to draft so it’s off the docket until it is ready, I overestimated the ease.

as a draft you will not run all tests so as it is at its advanced stage, let's keep it as full PR just add `wip` in the title

**Review by SkafteNicki — APPROVED** [review]

**Comment by SkafteNicki** [comment]
@ndrwrbgs the error only happens for older versions of PyTorch. The problem is that `torch.where`:
```python
torch.where(condition, input, other, *, out=None)
```
expects `other` and `input` to have same dtype in older versions of Pytorch whereas in newer versions of Pytorch it will automatically upcast to the highest precision. So we most likely need to explicit cast `zero_division` to the same dtype as `num/denom`

**Comment by ndrwrbgs** [comment]
Thank you! Promise I’ll get it fixed, I opened this just before visiting
family and have gotten busy 😊.

File: src/torchmetrics/utilities/compute.py [hunk 1]
```diff
@@ -56,8 +56,9 @@ def _safe_divide(num: Tensor, denom: Tensor, zero_division: float = 0.0) -> Tens
     """
     num = num if num.is_floating_point() else num.float()
     denom = denom if denom.is_floating_point() else denom.float()
-    zero_division = torch.tensor(zero_division).float().to(num.device)
-    return torch.where(denom != 0, num / denom, zero_division)
+    zero_division_tensor = torch.tensor(zero_division, dtype=num.dtype).to(num.device, non_blocking=True)
+
```
Comment (ndrwrbgs) [inline src/torchmetrics/utilities/compute.py:60]:
```suggestion
```
Reduce diff

File: src/torchmetrics/utilities/compute.py [hunk 2]
```diff
@@ -56,8 +56,8 @@ def _safe_divide(num: Tensor, denom: Tensor, zero_division: float = 0.0) -> Tens
     """
     num = num if num.is_floating_point() else num.float()
     denom = denom if denom.is_floating_point() else denom.float()
-    zero_division = torch.tensor(zero_division).float().to(num.device)
-    return torch.where(denom != 0, num / denom, zero_division)
+    zero_division_tensor = torch.tensor(zero_division, dtype=num.dtype).to(num.device, non_blocking=True)
```
Comment (ndrwrbgs) [inline src/torchmetrics/utilities/compute.py:59]: Renamed `zero_division` due to pylance complaining about re-typing the variable (as it was declared as `float` in the signature but is being assigned a `torch.Tensor`.

If so desired, alternatives include...
 - not rename it
 - retype the signature to Union[float, torch.Tensor]

**Comment by ndrwrbgs** [comment]
Thank you for your patience!

-----

I noticed errors like below in [this Azure job](https://dev.azure.com/Lightning-AI/Metrics/_build?definitionId=54&_a=summary):

```
error:  cannot create tests/_cache-references/tests/_cache-references/.unittests.retrieval.helpers._compute_sklearn_metric_276bbfaa62e3d293c69b71ebae3d3e4a266434eef12c60faa590f4111afbb6f7
        No space left on device
```

Here's an example both from my build and another before to confirm it's not introduced with this PR:
 - [build 1](https://dev.azure.com/Lightning-AI/Metrics/_build/results?buildId=210305&view=logs&j=95985737-7c02-5ecd-7847-688de4203f7e&t=2988db33-f785-5fa7-2e6d-77ab2e1977a7)
 - [build 2](https://dev.azure.com/Lightning-AI/Metrics/_build/results?buildId=210218&view=logs&j=95985737-7c02-5ecd-7847-688de4203f7e&t=2988db33-f785-5fa7-2e6d-77ab2e1977a7)

🤔 I'm not sure whether it has any observable impact or not.

**Comment by ndrwrbgs** [comment]
cc @SkafteNicki Another failure on the Windows 2.4.0 torch.Unique: https://github.com/Lightning-AI/torchmetrics/actions/runs/10150851419/job/28068885801?pr=2640

@Borda the other failures appear to be unrelated to the change, but I'm not seeing any rerun button available on my end. Is that something you're able to to with write permissions?
